### PR TITLE
Add "use exports { }" directive in es2015-modules-commonjs. Closes #6449

### LINKED ISSUE #6449: Add new pragma to transform-es2015-modules
image

**Open Question**: Should we make this default behavior, or put it behind a flag? If it is indeed on its way to being a part of node core, it makes sense to just make it default to `on`.

### PR DESCRIPTION
**Pull request by DrewML**
| Q                        | A 
| ------------------------ | ---
| Fixed Issues?            | Fixes #6449
| Patch: Bug Fix?          | No
| Major: Breaking Change?  | No
| Minor: New Feature?      | Yes
| Tests Added + Pass?      | No + yes
| Documentation PR         | No
| Any Dependency Changes?  | No

Makes the change as described in #6449 to the `transform-es2015-modules-commonjs` plugin.

I do have a 2 follow-up questions for @bmeck, though. 

1. What do you want us to do when `[[ExportName]]` would be `null` (`export * from 'mod'`)? Seems like you still need some hint of the exported names that would be coming from the module request. We're not able to handle that within Babel (operates on a single file at a time, no dep graph). Would it work if we extend the directive to reference which module requests had a `*` export?

2. [You mentioned that](https://github.com/bmeck/node/blob/named-export-core/doc/api/esm.md#use-exports-):
    > The properties in the prologue cannot be redefined to be getters or setters.

   Is this only a concern when _redefining_? Just want to make sure we're ok if they start as getters or setters at the time you read the directive.

**Notes**
- I considered adding a flag to disable this, but I can't think of a good reason to not just have it on all the time. I can't imagine there are tools that choke on unrecognized directives (and that would be reading compiled code). The only thing that is making me hesitate is this line from the spec:

    > If an appropriate notification mechanism exists, an implementation should issue a warning if it encounters in a Directive Prologue an ExpressionStatement that is not a Use Strict Directive and which does not have a meaning defined by the implementation.

### CODE REVIEW

File: packages/babel-plugin-transform-es2015-modules-commonjs/src/index.js [hunk 1]
```diff
@@ -134,6 +134,21 @@ export default function({ types: t, template }) {
             noInterop,
           });
 
+          let exportsList = [];
```
Comment (DrewML) [inline packages/babel-plugin-transform-es2015-modules-commonjs/src/index.js]: Some style guides would forbid this, but I felt like it was more readable than my initial implementations using `map` or `reduce` and flattening. Let me know if a change is desired

**Comment by bmeck** [comment]
> I considered adding a flag to disable this, but I can't think of a good reason to not just have it on all the time. I can't imagine there are tools that choke on unrecognized directives (and that would be reading compiled code). The only thing that is making me hesitate is this line from the spec:

I'm in talks w/ webpack which should be the main place that isn't node using CJS these days.

> What do you want us to do when [[ExportName]] would be null (export * from 'mod')? Seems like you still need some hint of the exported names that would be coming from the module request. We're not able to handle that within Babel (operates on a single file at a time, no dep graph). Would it work if we extend the directive to reference which module requests had a * export?

We could extend it, but there might be a mismatch since the ESM couldn't defer to the `exports` object.

```js
"use exports {foo}, * from 'react'"
```

Would have Node generate the ~ ESM facade of:

```js
var foo;
export {foo};
export * from 'react';
```

While the CJS compiled form would be something like:

```js
var react = require('react');
Object.defineProperties(exports, {
  ... key of react ... : { get() { ... } }
});
```

Which I think is fine, but need to think on it. Especially since the specifier in ESM facade should be using require to resolve instead of ESM to have consistency.

> Is this only a concern when redefining? Just want to make sure we're ok if they start as getters or setters at the time you read the directive.

The properties can never be getter/setters. Exported properties names start with an initially defined property descriptor:

```js
{
  value: undefined
  writable: true,
  configurable: true
  enumerable: true
}
```

With a Proxy guard to prevent redefining those properties to incompatible descriptors (getter/setters).

**Comment by bmeck** [comment]
Note that with the `export * from` I don't include the list of identifiers from the dependency in the `{}`s so they would not have a Proxy guard.

**Comment by DrewML** [comment]
Should note that we also use a getter right now for `export {x} from "mod";` to maintain live bindings. Not sure how we'd handle that in this implementation if we can only do simple assignment.

**Comment by bmeck** [comment]
@DrewML the proxy invariants are different for indirect exports vs direct exports, we can and should back those by non-configurable getters

File: packages/babel-plugin-transform-es2015-modules-commonjs/src/index.js (hunk 1 above)
Comment (Kovensky) [reply packages/babel-plugin-transform-es2015-modules-commonjs/src/index.js]: Hmm, I don't find this too hard to read......
```js
const exportsList = [].concat(
  ...Object.values(meta.local).map({ names } => names),
  ...Array.from(meta.source.values(), source => Array.from(source.reexports.keys()))
)
```
...but I'd definitely want comments if written like this.

**Comment by bmeck** [comment]
going to be writing up more spec like text here, lots of edits in next few

**Comment by guybedford** [comment]
These interop directions look really amazing!

I would suggest to the PR to ensure __esModule is included in the list, so that the __esModule interop can continue to work between CommonJS modules that don't use this feature and CommonJS modules that do. Unless I've missed something in the logic, but I think that would work out nicely.

**Comment by guybedford** [comment]
To try and clarify the breaking case above:

* Publish a library CommonJS to npm that uses this "use exports {...}" feature
* Someone authors a CommonJS package that it also transpiled with __esModule interop and depends on the library.
* It works fine in CommonJS with modules disabled, but as soon as modules are enabled, the interop breaks.

Hence, __esModule being exported would correct that. Not correcting this may add a viral property here, but best not to think in those terms!

**Comment by bmeck** [comment]
@guybedford the target audience for this is tools, so they could always enforce adding that.

**Comment by guybedford** [comment]
@bmeck certainly other tools don't have to use it, but for Babel specifically here it is important to consistent interop without breakage.

**Comment by ljharb** [comment]
100% agree with @guybedford; `__esModule` has to persist for back compat.

**Comment by benjamn** [comment]
Just asked this question in the original issue: https://github.com/babel/babel/issues/6449#issuecomment-337434223

Please don't abandon the `__esModule` convention lightly (it's not perfect, but it is a widely-adopted convention), and please think twice about introducing an unspecified directive that duplicates part of its semantics. This proposal strikes me as incredibly ad-hoc.

Neither `"use module"` nor `export {}` have gained acceptance so far, and both of those ideas seem considerably simpler than this one.

**Comment by ljharb** [comment]
@benjamn both of those are proposals to handle ambiguous parsing cases; this is unrelated to that - it's for statically declaring exported names in CJS modules, to open up the possibility of ESM being able to named-import from them.

**Comment by azz** [comment]
Minor bikeshed, but why `"use exports ..."`? Why not just `"exports ..."`. The exports are _used_ by other modules, not the module itself.

**Comment by bmeck** [comment]
@azz I am not tied to any syntax within the pragma as long as capabilities stay the same. Current version is matching other directive prologues (`"use strict"`, `"use module"` RIP) that affect eval like adding the Proxy invariants here. I think it is fine to change.

**Comment by bmeck** [comment]
@benjamn This directive prologue is for systems consuming files and wishing to create an ESM facade to them. It is not meant to change the goal of a file, but it does introduce invariants to the `exports` which ensure the listed properties of `exports` stay in sync with the ESM namespace. The invariants are able to be stripped by an environment which does not wish to enforce them, but listing them and saying they will be enforced gives compilers knowledge of what they can/cannot do safely.

In essence we want ESM -> CJS -> ESM transpilation to preserve the list of named exports. 

Currently ESM -> CJS will take the entire module namespace and wrap it into a single `default` exports, which removes all named exports. This is done to allow creation of Module Records that can be statically [Instantiated](https://tc39.github.io/ecma262/#sec-moduledeclarationinstantiation) will a well defined shape prior to evaluation. Environments generating spec based Module Records and preserving order of evaluation cannot wait until the evaluation phase to link the module graph. The entire graph is linked prior to any evaluation.

This pragma would give a way for an environment like Node or Webpack to create a Module Record with the proper shape:

```js
// original
export let foo = 1;
// import {foo} from 'original' works
```

```js
// compiled
"use exports {foo}"
exports.foo = 1;
// import {foo} from 'compiled' works
```

**Comment by benjamn** [comment]
I don't object to Babel generating something like this, because it certainly could be useful information for other tools to consume. I'm more worried about Node making this syntax meaningful, since tooling conventions that Node adopts tend to become de facto standards, even if they have fatal flaws and will be painful to deprecate.

Here are some questions I would love to see answered before I can get behind this proposal:
1. How does this proposal help with CJS modules whose compilation you don't control, such as those installed from npm? Would it not be preferable to put the metadata outside the module, a la TypeScript declaration files?
2. What exactly happens if the directive syntax has a parse error? How does the developer find out? Is the directive just ignored?
3. What exactly happens if the declared exports disagree with the shape of the `module.exports` object?
4. Is this information available at runtime as well? You could imagine (instead of this proposal) expanding the `exports.__esModule` object to be (1) still truthy, and both (2) statically and (3) dynamically analyzable:
```js
// Object literal syntax is required for static checking to work:
exports.__esModule = {
  // These values could provide more useful information, even if some tools ignore them.
  // Maybe source module identifier strings?
  foo: 1,
  default: 1,
};
```
5. Can you think of any past precedents where Node gave special meaning to Node-specific metadata that had not been standardized? For example, `package.json` certainly counts, but `"use strict"` is part of ECMAScript, so not Node-specific. Can we learn anything from those examples?

I'm sorry my initial reaction to this has been so negative, but I think this warrants serious discussion, and I don't feel like this PR thread represents anything like a quorum of the stakeholders.

**Comment by jdalton** [comment]
> but I think this warrants serious discussion, and I don't feel like this PR thread represents anything like a quorum of the stakeholders.

The fact that this is even in PR form now kind of worries me. If feels like it's rushing to completion and that everything is settled *(no room for discussion)*.

This isn't just a compile time thing though. This is something Node is suggesting users who write CJS but want their modules consumed with named exports, more than `default`, should write. That means this isn't something we can just shrug and go "well let transpilers handle it". It's something that devs will have to manage *(kinda gross)*.

**Comment by jdalton** [comment]
Is this discussion splintered in other areas (a Node-eps issue or other threads, maybe webpack?). Maybe a central Node-land issue is a better place than a primed Babel PR.

**Comment by jdalton** [comment]
I'm also concerned with the performance of this approach. From [Bradley's PRs](https://github.com/bmeck/node/commits/named-export-core) it looks like performance may be an issue. With Node ESM slower than CJS equivs do we know how much is this going to bog things down on Node's side?

**Comment by bmeck** [comment]
@benjamn 

> 1. How does this proposal help with CJS modules whose compilation you don't control, such as those installed from npm? Would it not be preferable to put the metadata outside the module, a la TypeScript declaration files?

It intentionally does not, you do not control those files, vendor them using bundledModules or into a directory. This has affects on what can be done to `module.exports` since the value cannot be replaced and `exports` which has a Proxy enforcing invariants on it. It is not safe to add to 3rd party code. Vendor it and make it 1st party.

> 2. What exactly happens if the directive syntax has a parse error? How does the developer find out? Is the directive just ignored?

In environments not supporting the directive it will be ignored, just as strict mode was when it was first introduced. In environments supporting it, we should clarify what happens. An early error I would suggest.

> 3. What exactly happens if the declared exports disagree with the shape of the module.exports object? When would that agreement be checked?

`module.exports` cannot be re-assigned. This is a non issue since the proxy would enforce the invariants described above, including not being able to remove the properties listed to be exposed as ESM exports.

> 4. Is this information available at runtime as well? 

Up to the environment.

> You could imagine (instead of this proposal) expanding the exports.__esModule object

This is not statically parsable and therefore not a viable path towards allowing any sort of Module Record in ECMA262 to be Instatiated properly before evaluation.

> Can you think of any past precedents where Node gave special meaning to Node-specific metadata that had not been standardized? For example, package.json certainly counts, but "use strict" is part of ECMAScript, so not Node-specific. Can we learn anything from those examples?

`NODE_PATH` is the main one. Interestingly some people still consider `package.json#main` a wart. Per file data via file extension are also used. `NODE_PATH` was removed due to affects and reliance upon it affecting modules not prepared for it. A lack of ability to guard against strange `NODE_PATH` was probably what I see as the biggest thing to take away.

Since this is per file and does not affect other files, I think this is the only example really of data like this.

@jdalton 

> It feels like everything is settled (I know a lot is still up in the air).

Correct, this is all up in the air.

> This isn't just a compile-time thing. This is something Node is suggesting users, who write CJS but want their modules consumed with named exports more than default, should write. That means this isn't something we can just shrug and go "Well, let transpilers handle it". It's something that devs will have to manage and write too (kinda gross).

If you can think of a viable alternative I am fine with whatever. But it must be statically analyzable in order to generate the Module Records. It cannot require evaluation.

> From Bradley's commits it looks like performance may be an issue. With Node ESM slower than CJS equivs do we know how much further this is going to bog things down on Node's side?

I need benchmarks to understand this comment. It does slow down CJS using this, but it does not slow down ESM (since ESM doesn't need to add the invariant checks to the Module Record).

> Is this discussion occurring in other places, a Node-eps issue or webpack issue?
A central Node-land issue is probably a better place than a primed Babel PR.

We can move all of this to https://github.com/babel/babel/issues/6449 ?

**Comment by benjamn** [comment]
> This is not statically parsable and therefore not a viable path towards allowing any sort of Module Record in ECMA262 to be Instatiated properly before evaluation.

I'm not suggesting we support any ol' runtime strategy for dynamically defining `exports.__esModule`. It would have to be an assignment to a member expression whose property is `__esModule` (knowing if the object is `exports` would be nice, but doesn't seem vital), with an object literal on the right-hand side whose keys are statically parseable. Otherwise, it wouldn't have any special meaning.

This is no more or less statically analyzable (in the absolute sense, ignoring relative difficulty) than an arbitrary string literal expression statement that happens to sometimes endow the module with special meaning in certain environments, provided you follow the right rules.

**Comment by jdalton** [comment]
> I need benchmarks to understand this comment. It does slow down CJS using this, but it does not slow down ESM (since ESM doesn't need to add the invariant checks to the Module Record).

My comment about ESM being slower is [what I've observed](https://medium.com/web-on-the-edge/es-modules-in-node-today-32cff914e4b#c08f). My question was; given that ESM is slower how much more are things going to take a hit? I had thought since named exports are only a concern in ESM, that the scenario would consuming CJS in ESM, but you're saying ALL CJS would take a perf hit?

**Comment by bmeck** [comment]
@jdalton all CJS takes a minor hit due to checking for directive prologue, it isn't significant enough to get a clear read in any of my naive tests. There isn't a need for full AST parsing and there is a fairly early bailout condition for this check.

CJS using this pragma will take a hit on reading or assigning properties to `exports` due to it being a Proxy. Newer V8s make it roughly on par with a getter/setter on an object that has been forced into dictionary mode.

ESM using the named exports will not have any difference in speed from ESM<->ESM once the variables are setup since the bindings are shared and don't invoke any sort of Proxy mechanism.

> My comment about ESM being slower is what I've observed

This is not inherent to the spec, but neither V8 nor Node implementation has really been optimized; being on par with userland makes sense while correctness is still the overbearing concern. There is an incredible amount of parser/GC waste currently and it is still single threaded for now. That won't be changing for a while.

**Comment by jdalton** [comment]
I'm confused. Your first reply was *"It does slow down CJS using this"* but now your reply is diminishing the that bit. So is there a noticeable perf hit or not? Have you checked how this impacts runtime or load performance? Knowing the perf constraints can help inform implementation and also open folks to alternatives.

**Comment by bmeck** [comment]
It should be noted that even with the pragma; if you are consuming through ESM, there is no penalty after startup.

**Comment by jdalton** [comment]
> If you do not use the pragma, no.

At the moment the Babel PR looks to be adding the pragma for everything even when only exporting a default. Now knowing that there's a perf hit only with the pragma use Babel could modify it's inclusion of the pragma to only happen when exporting more than `default`.

> If you use the pragma, only when accessing / assigning to the Proxy.

So a proxy is only used for CJS that include the pragma?

> It should be noted that even with the pragma; if you are consuming through ESM, there is no penalty after startup.

Cool! Startup/load perf is something I've seen devs concerned with.

**Comment by jdalton** [comment]
You could also optimize Node's handling of the pragma to punt if its only exporting a "default" too:
```
"use export { default }"

**Comment by bmeck** [comment]
> So a proxy is only used for CJS that include the pragma?

Yes.

**Comment by jdalton** [comment]
Ok so prelim perf results.
I measured loading `lodash-es` (643 modules) babel converted *(so CJS)* with pragmas from this PR.

Something like:
<details><summary>bench</summary>
<pre><code>
#!/usr/bin/env bash
start=$(gdate +%s.%3N)
../node/out/Release/node ./node_modules/lodash-babel/lodash.js
dur=$(echo "$(gdate +%s.%3N) - $start" | bc)
printf "Execution time: %.3f seconds" $dur
</code></pre>
</details>
<br>

Node 8.7.0 without `--experimental-modules`:
```
Execution time: 0.267 seconds%
```
Node 8.7.0 with `--experimental-modules`:
```
Execution time: 0.450 seconds%
```
Based on the `named-export-core` branch *(pragma parsing+proxy)*:
```
Execution time: 0.682 seconds%
```
So more than a 2.5x perf hit on load time.

**Comment by bmeck** [comment]
Since it does create Module Records on my branch it makes more sense to compare it with the flagged ESM than regular CJS Node.

**Comment by jdalton** [comment]
>
Ah, that's worse
```
Execution time: 0.851 seconds%
```
So a ~3.2x perf hit on startup/load time 📉

**Comment by bmeck** [comment]
`0.450` vs `0.851`?

**Comment by jdalton** [comment]
There's a couple of views here. My `~n.nX` measurements are comparing CJS in Node today so the Node v8.70 without `--experimental-modules` of `0.267` to other runs. If you only care about the `--experimental-modules` to other `--experimental-modules` run then ya go from `0.450`.

**Comment by bmeck** [comment]
I will always run against `--experimental-modules` since thats the target for this feature.

**Comment by loganfsmyth** [comment]
I'm triaging old PR on Babel. Since this has been sitting for ages, I'm going to close it. The conversation can always continue or we can reopen later.

**Pull request closed without merging**